scipy/optimize/_trustregion_iterative.py

```diff
@@ -0,0 +1,470 @@
+        self.CLOSE_TO_ZERO = self.dimension*self.EPS*self.hess_inf
+
+    @property
+    def dimension(self):
+        """Hessian inf norm"""
+        if self._dimension is None:
+            self._dimension = len(self.hess)
+        return self._dimension
+
+    @property
+    def hess_gershgorin_ub(self):
+        """Gershgorin upper bound defined on p.19 from [1]
+        computed for the hessian"""
+        if self._hess_gershgorin_ub is None:
+            n = self.hess.shape[0]
+
+            # Compute H[i,i] + sum(abs(H[i, j]), i!=j)
+            maximum = -np.Inf
+            for i in range(n):
+                summation = self.hess[i, i]
+                for j in range(n):
+                    summation += abs(self.hess[i, j]) if i != j else 0
+
+                if summation > maximum:
+                    maximum = summation
+
+            self._hess_gershgorin_ub = maximum
+        return self._hess_gershgorin_ub
+
+    @property
+    def hess_gershgorin_lb(self):
+        """Gershgorin lower bound defined on p.19 from [1]
+        computed for the hessian"""
+        if self._hess_gershgorin_lb is None:
+            n = self.hess.shape[0]
+
+            # Compute -H[i,i] + sum(abs(H[i, j]), i!=j)
+            maximum = -np.Inf
+            for i in range(n):
+                summation = -self.hess[i, i]
+                for j in range(n):
+                    summation += abs(self.hess[i, j]) if i != j else 0
+
+                if summation > maximum:
+                    maximum = summation
+
+            self._hess_gershgorin_lb = -maximum
+        return self._hess_gershgorin_lb
+
+    @property
+    def hess_inf(self):
+        """Hessian inf norm"""
+        if self._hess_inf is None:
+            self._hess_inf = norm(self.hess, np.Inf)
+        return self._hess_inf
+
+    @property
+    def hess_fro(self):
+        """Hessian frobenius norm"""
+        if self._hess_frob is None:
+            self._hess_frob = norm(self.hess, 'fro')
+        return self._hess_frob
+
+    def _initial_values(self, trust_radius):
+        """Given a trust radius, return a good initial guess for
+        the damping factor, the lower bound and the upper bound.
+        The values were chosen accordingly to the guidelines on
+        section 7.3.8 (p. 192) from [1].
+        """
+
+        # Upper bound
+        ub = max(0, self.jac_mag/trust_radius+min(-self.hess_gershgorin_lb,
+                                                  self.hess_fro,
+                                                  self.hess_inf))
+
+        # Lower bound
+        lb = max(0, -min(self.hess.diagonal()),
+                 self.jac_mag/trust_radius-min(self.hess_gershgorin_ub,
+                                               self.hess_fro,
+                                               self.hess_inf))
+
+        # Improve bounds with previous info
+        if trust_radius < self.previous_trust_radius:
+            lb = max(self.lb, lb)
+
```

Comment:
Look here to see the `if` the algorithm will enter case it have already been executed before. The first time the algorithm will not enter in this section because at first `self.previous_trust_radius = -1`